File: frontend/src/utils/genetics.ts
import type { JetMorph, MetricGenome, Palette, Phenotype, ViewConfig } from '../types/genome'
import type { Base, Feature, MetricSpec } from '../types/spec'
import { availableFeaturesForSplice, FEATURE_PARAM_DEFAULTS } from '../types/spec'
import { JET_MORPH_NAMES, PALETTE_NAMES } from '../data/palettes'
import { compileSpec } from './compileSpec'

let uidCounter = 100

function nextId(): string {
  return `g-${String(++uidCounter).padStart(4, '0')}`
}

function gaussian(stdDev: number): number {
  const u1 = Math.random()
  const u2 = Math.random()
  return stdDev * Math.sqrt(-2 * Math.log(u1 + 1e-10)) * Math.cos(2 * Math.PI * u2)
}

const clamp01  = (v: number): number => Math.max(0, Math.min(1, v))
const clampPos = (v: number, eps = 0.05): number => Math.max(eps, v)

// ── Spec mutation: drift, splice, prune, transitionBase ──────────────────────

// Magnitude parameters that must stay positive for the metric to be physical
// (mass M, charge magnitudes Q/Qa, spin |a|, regularization length scales).
// Everything else (Λ, c3, c4, cy) is allowed to drift through zero — that's
// where the "free sign flip" mutation comes from.
const POSITIVE_ONLY = new Set(['M', 'Q', 'Qa', 'a', 'ell', 'eb', 'Ly'])

function driftParameters(params: Record<string, number>, std: number): Record<string, number> {
  const out: Record<string, number> = {}
  for (const [k, v] of Object.entries(params)) {
    const drifted = v + gaussian(std) * (Math.abs(v) + 0.05)
    out[k] = POSITIVE_ONLY.has(k) ? clampPos(drifted) : drifted
  }
  return out
}

// Splice: add a new feature to the spec, with default parameter values.
// Returns spec unchanged if the base doesn't accept any new features or
// all available features are already present.
function spliceFeature(spec: MetricSpec): MetricSpec {
  const candidates = availableFeaturesForSplice(spec)
  if (candidates.length === 0) return spec

  const kind = candidates[Math.floor(Math.random() * candidates.length)]
  const defaults = FEATURE_PARAM_DEFAULTS[kind]
  return {
    ...spec,
    features:   [...spec.features, { kind } as Feature],
    parameters: { ...spec.parameters, ...defaults },
  }
}

// Prune: drop a random feature and its parameters.
function pruneFeature(spec: MetricSpec): MetricSpec {
  if (spec.features.length === 0) return spec
  const idx = Math.floor(Math.random() * spec.features.length)
  const removed = spec.features[idx]
  const removedParams = FEATURE_PARAM_DEFAULTS[removed.kind]
  const newParams = { ...spec.parameters }
  for (const k of Object.keys(removedParams)) delete newParams[k]
  return {
    ...spec,
    features:   spec.features.filter((_, i) => i !== idx),
    parameters: newParams,
  }
}

// Rare base transition: Schwarzschild ↔ Kerr. Drops features (since
// Kerr-base features aren't implemented) and seeds the new parameter set
// with sensible defaults pulled from the previous spec where applicable.
function transitionBase(spec: MetricSpec): MetricSpec {
  const M = spec.parameters.M ?? 1.0
  if (spec.base.kind === 'schwarzschild') {
    return {
      base:       { kind: 'kerr' },
      features:   [],
      parameters: { M, a: 0.40 },
    }
  }
  return {
    base:       { kind: 'schwarzschild' },
    features:   [],
    parameters: { M },
  }
}

function mutateSpec(spec: MetricSpec, std: number, rerollRate: number): MetricSpec {
  let out: MetricSpec = { ...spec, parameters: driftParameters(spec.parameters, std) }
  // Splice / prune / transition are independent low-probability events.
  // Rates are tied to the mutation slider.
  const spliceProb     = rerollRate * 0.45
  const pruneProb      = rerollRate * 0.30
  const transitionProb = rerollRate * 0.10
  if (Math.random() < spliceProb)     out = spliceFeature(out)
  if (Math.random() < pruneProb)      out = pruneFeature(out)
  if (Math.random() < transitionProb) out = transitionBase(out)
  return out
}

// ── View / phenotype mutation (continuous + discrete) ────────────────────────

function mutateView(v: ViewConfig, std: number): ViewConfig {
  return {
    inclination:    clamp01(v.inclination    + gaussian(std)),
    diskBrightness: clamp01(v.diskBrightness + gaussian(std)),
    diskRadius:     clamp01(v.diskRadius     + gaussian(std)),
    jetStrength:    clamp01(v.jetStrength    + gaussian(std)),
  }
}

function mutatePhenotype(p: Phenotype, std: number, rerollRate: number): Phenotype {
  return {
    palette:        Math.random() < rerollRate ? randomPalette()  : p.palette,
    jetMorph:       Math.random() < rerollRate ? randomJetMorph() : p.jetMorph,
    starSeed:       Math.random() < rerollRate ? Math.floor(Math.random() * 1000) : p.starSeed,
    diskTurbulence: clamp01(p.diskTurbulence + gaussian(std)),
  }
}

function randomPalette(): Palette {
  return PALETTE_NAMES[Math.floor(Math.random() * PALETTE_NAMES.length)]
}

function randomJetMorph(): JetMorph {
  return JET_MORPH_NAMES[Math.floor(Math.random() * JET_MORPH_NAMES.length)]
}

// ── Crossover ────────────────────────────────────────────────────────────────

function pickField<T, K extends keyof T>(a: T, b: T, k: K): T[K] {
  return Math.random() < 0.5 ? a[k] : b[k]
}

// Same-base parents may have different feature sets; we union the sets and
// for shared parameters pick from one parent at random. Different-base
// parents fall back to inheriting one parent's spec wholesale.
function crossSpecs(a: MetricSpec, b: MetricSpec): MetricSpec {
  if (a.base.kind !== b.base.kind) {
    return Math.random() < 0.5 ? a : b
  }
  const base: Base = { kind: a.base.kind }

  // Union of features by kind, each with 50/50 chance of inclusion if it
  // appears in just one parent and 100% if it appears in both.
  const aKinds = new Set(a.features.map((f) => f.kind))
  const bKinds = new Set(b.features.map((f) => f.kind))
  const allKinds = new Set([...aKinds, ...bKinds])
  const features: Feature[] = []
  for (const kind of allKinds) {
    const inBoth = aKinds.has(kind) && bKinds.has(kind)
    if (inBoth || Math.random() < 0.5) {
      features.push({ kind } as Feature)
    }
  }

  // Build the parameter set: each parameter's value picked from whichever
  // parent had it (or both, in which case 50/50).
  const allParamKeys = new Set([...Object.keys(a.parameters), ...Object.keys(b.parameters)])
  const parameters: Record<string, number> = {}
  for (const k of allParamKeys) {
    if (k in a.parameters && k in b.parameters) {
      parameters[k] = pickField(a.parameters, b.parameters, k)
    } else if (k in a.parameters) {
      parameters[k] = a.parameters[k]
    } else {
      parameters[k] = b.parameters[k]
    }
  }

  // Drop parameters that aren't needed and fill in any missing ones with
  // sensible defaults (a child can need a param that neither parent had if
  // the union of features grew).
  const required = new Map<string, number>()
  if (base.kind === 'kerr')          { required.set('M', 1.0); required.set('a', 0.4) }
  if (base.kind === 'schwarzschild') { required.set('M', 1.0) }
  for (const f of features) {
    for (const [k, v] of Object.entries(FEATURE_PARAM_DEFAULTS[f.kind])) {
      required.set(k, v)
    }
  }
  const filtered: Record<string, number> = {}
  for (const [k, fallback] of required) {
    filtered[k] = k in parameters ? parameters[k] : fallback
  }

  return { base, features, parameters: filtered }
}

// ── Genome mutation / crossover ─────────────────────────────────────────────

function buildChild(spec: MetricSpec, view: ViewConfig, phenotype: Phenotype, baseName: string): MetricGenome {
  const id = nextId()
  const compiled = compileSpec(spec)
  return {
    id,
    name:          `${baseName}-${id.slice(-3)}`,
    spec,
    metric:        compiled.metric,
    displayMetric: compiled.displayMetric,
    view,
    phenotype,
  }
}

function mutate(g: MetricGenome, std: number, rerollRate: number): MetricGenome {
  const baseName = g.name.split('-')[0]
  return buildChild(
    mutateSpec(g.spec, std, rerollRate),
    mutateView(g.view, std),
    mutatePhenotype(g.phenotype, std, rerollRate),
    baseName,
  )
}

function crossover(a: MetricGenome, b: MetricGenome): MetricGenome {
  const view: ViewConfig = {
    inclination:    pickField(a.view, b.view, 'inclination'),
    diskBrightness: pickField(a.view, b.view, 'diskBrightness'),
    diskRadius:     pickField(a.view, b.view, 'diskRadius'),
    jetStrength:    pickField(a.view, b.view, 'jetStrength'),
  }
  const phenotype: Phenotype = {
    palette:        pickField(a.phenotype, b.phenotype, 'palette'),
    jetMorph:       pickField(a.phenotype, b.phenotype, 'jetMorph'),
    starSeed:       pickField(a.phenotype, b.phenotype, 'starSeed'),
    diskTurbulence: pickField(a.phenotype, b.phenotype, 'diskTurbulence'),
  }
  return buildChild(crossSpecs(a.spec, b.spec), view, phenotype, a.name.split('-')[0])
}

// `strength` ∈ [0, 1]: dilates Gaussian σ and reroll/splice/prune probabilities.
export function breedGeneration(
  parents: MetricGenome[],
  strength = 0.5,
  targetCount = 6,
): MetricGenome[] {
  const s = Math.max(0, Math.min(1, strength))
  const std        = s * 0.45
  const rerollRate = s * 0.85

  if (parents.length === 0) {
    throw new Error('breedGeneration requires at least 1 parent')
  }
  if (parents.length === 1) {
    return Array.from({ length: targetCount }, () => mutate(parents[0], std, rerollRate))
  }
  return Array.from({ length: targetCount }, () => {
    const idxA = Math.floor(Math.random() * parents.length)
    let idxB = Math.floor(Math.random() * (parents.length - 1))
    if (idxB >= idxA) idxB++
    // Crossover already mixes from two parents; soften the post-mutation.
    return mutate(crossover(parents[idxA], parents[idxB]), std * 0.5, rerollRate * 0.5)
  })
}

File: services/breeding-service/app/genetics.py
"""
Python port of frontend/src/utils/genetics.ts.

The TS source is authoritative for the algorithm shape; this module
preserves the same control flow and probability constants. RNG output
is *not* bit-identical to TS Math.random() — different generators —
but the structural invariants (target_count, base preservation under
crossover, splice avoiding duplicates, etc.) are tested in
tests/test_genetics.py.

Operates on MetricSpec (the JSON-able recipe). The compiled Metric AST
stays client-side in compileSpec.ts; this service never sees it.
"""

import random
from uuid import uuid4

from app.schemas import (
    Feature,
    Genome,
    KerrBase,
    MetricSpec,
    Phenotype,
    SchwarzschildBase,
    ViewConfig,
)
from app.spec_meta import (
    FEATURE_PARAM_DEFAULTS,
    JET_MORPH_NAMES,
    PALETTE_NAMES,
    available_features_for_splice,
)

# Magnitude parameters that must stay positive for the metric to be physical.
# Everything else (Lambda, c3, c4, cy) is allowed to drift through zero —
# that's where free sign flips come from.
POSITIVE_ONLY = frozenset({"M", "Q", "Qa", "a", "ell", "eb", "Ly"})


def _gaussian(std_dev: float) -> float:
    return random.gauss(0.0, std_dev)


def _clamp01(v: float) -> float:
    return max(0.0, min(1.0, v))


def _clamp_pos(v: float, eps: float = 0.05) -> float:
    return max(eps, v)


# ── Spec mutation: drift / splice / prune / transition ─────────────────────


def _drift_parameters(params: dict[str, float], std: float) -> dict[str, float]:
    out: dict[str, float] = {}
    for k, v in params.items():
        drifted = v + _gaussian(std) * (abs(v) + 0.05)
        out[k] = _clamp_pos(drifted) if k in POSITIVE_ONLY else drifted
    return out


def _splice_feature(spec: MetricSpec) -> MetricSpec:
    candidates = available_features_for_splice(spec)
    if not candidates:
        return spec
    kind = random.choice(candidates)
    defaults = FEATURE_PARAM_DEFAULTS[kind]
    return MetricSpec(
        base=spec.base,
        features=[*spec.features, Feature(kind=kind)],
        parameters={**spec.parameters, **defaults},
    )


def _prune_feature(spec: MetricSpec) -> MetricSpec:
    if not spec.features:
        return spec
    idx = random.randrange(len(spec.features))
    removed = spec.features[idx]
    removed_params = FEATURE_PARAM_DEFAULTS[removed.kind]
    new_params = {k: v for k, v in spec.parameters.items() if k not in removed_params}
    new_features = [f for i, f in enumerate(spec.features) if i != idx]
    return MetricSpec(base=spec.base, features=new_features, parameters=new_params)


def _transition_base(spec: MetricSpec) -> MetricSpec:
    M = spec.parameters.get("M", 1.0)
    if spec.base.kind == "schwarzschild":
        return MetricSpec(
            base=KerrBase(kind="kerr"),
            features=[],
            parameters={"M": M, "a": 0.40},
        )
    return MetricSpec(
        base=SchwarzschildBase(kind="schwarzschild"),
        features=[],
        parameters={"M": M},
    )


def _mutate_spec(spec: MetricSpec, std: float, reroll_rate: float) -> MetricSpec:
    out = MetricSpec(
        base=spec.base,
        features=spec.features,
        parameters=_drift_parameters(spec.parameters, std),
    )
    if random.random() < reroll_rate * 0.45:
        out = _splice_feature(out)
    if random.random() < reroll_rate * 0.30:
        out = _prune_feature(out)
    if random.random() < reroll_rate * 0.10:
        out = _transition_base(out)
    return out


# ── View / phenotype mutation ──────────────────────────────────────────────


def _mutate_view(v: ViewConfig, std: float) -> ViewConfig:
    return ViewConfig(
        inclination=_clamp01(v.inclination + _gaussian(std)),
        diskBrightness=_clamp01(v.diskBrightness + _gaussian(std)),
        diskRadius=_clamp01(v.diskRadius + _gaussian(std)),
        jetStrength=_clamp01(v.jetStrength + _gaussian(std)),
    )


def _mutate_phenotype(p: Phenotype, std: float, reroll_rate: float) -> Phenotype:
    return Phenotype(
        palette=random.choice(PALETTE_NAMES) if random.random() < reroll_rate else p.palette,
        jetMorph=random.choice(JET_MORPH_NAMES) if random.random() < reroll_rate else p.jetMorph,
        starSeed=random.randint(0, 999) if random.random() < reroll_rate else p.starSeed,
        diskTurbulence=_clamp01(p.diskTurbulence + _gaussian(std)),
    )


# ── Crossover ──────────────────────────────────────────────────────────────


def _cross_specs(a: MetricSpec, b: MetricSpec) -> MetricSpec:
    # Different bases: inherit one wholesale (can't usefully merge spaces).
    if a.base.kind != b.base.kind:
        return a if random.random() < 0.5 else b

    a_kinds = {f.kind for f in a.features}
    b_kinds = {f.kind for f in b.features}
    features: list[Feature] = []
    for kind in a_kinds | b_kinds:
        in_both = kind in a_kinds and kind in b_kinds
        if in_both or random.random() < 0.5:
            features.append(Feature(kind=kind))

    # Mix parameter values when both parents have the key, else inherit.
    parameters: dict[str, float] = {}
    for k in set(a.parameters) | set(b.parameters):
        if k in a.parameters and k in b.parameters:
            parameters[k] = a.parameters[k] if random.random() < 0.5 else b.parameters[k]
        elif k in a.parameters:
            parameters[k] = a.parameters[k]
        else:
            parameters[k] = b.parameters[k]

    # Filter to what the resulting (base + features) actually requires; fill
    # missing keys with defaults (a child's feature set may be a superset of
    # either parent's, so a required param may not be in `parameters`).
    required: dict[str, float] = {}
    if a.base.kind == "kerr":
        required["M"] = 1.0
        required["a"] = 0.4
    elif a.base.kind == "schwarzschild":
        required["M"] = 1.0
    for f in features:
        for k, default in FEATURE_PARAM_DEFAULTS[f.kind].items():
            required[k] = default

    filtered = {k: parameters.get(k, fallback) for k, fallback in required.items()}
    return MetricSpec(base=a.base, features=features, parameters=filtered)


def _crossover(a: Genome, b: Genome) -> Genome:
    view = ViewConfig(
        inclination=a.view.inclination if random.random() < 0.5 else b.view.inclination,
        diskBrightness=(
            a.view.diskBrightness if random.random() < 0.5 else b.view.diskBrightness
        ),
        diskRadius=a.view.diskRadius if random.random() < 0.5 else b.view.diskRadius,
        jetStrength=a.view.jetStrength if random.random() < 0.5 else b.view.jetStrength,
    )
    phenotype = Phenotype(
        palette=a.phenotype.palette if random.random() < 0.5 else b.phenotype.palette,
        jetMorph=a.phenotype.jetMorph if random.random() < 0.5 else b.phenotype.jetMorph,
        starSeed=a.phenotype.starSeed if random.random() < 0.5 else b.phenotype.starSeed,
        diskTurbulence=(
            a.phenotype.diskTurbulence
            if random.random() < 0.5
            else b.phenotype.diskTurbulence
        ),
    )
    base_name = a.name.split("-")[0]
    return _build_child(_cross_specs(a.spec, b.spec), view, phenotype, base_name)


# ── Top-level: build_child / mutate / breed_generation ─────────────────────


def _build_child(
    spec: MetricSpec, view: ViewConfig, phenotype: Phenotype, base_name: str
) -> Genome:
    new_id = uuid4()
    return Genome(
        id=new_id,
        name=f"{base_name}-{str(new_id)[-3:]}",
        spec=spec,
        view=view,
        phenotype=phenotype,
    )


def _mutate(g: Genome, std: float, reroll_rate: float) -> Genome:
    base_name = g.name.split("-")[0]
    return _build_child(
        _mutate_spec(g.spec, std, reroll_rate),
        _mutate_view(g.view, std),
        _mutate_phenotype(g.phenotype, std, reroll_rate),
        base_name,
    )


def breed_generation(
    parents: list[Genome], strength: float = 0.5, target_count: int = 6
) -> list[Genome]:
    """Produce `target_count` offspring from one or more parents.

    `strength` ∈ [0, 1] dilates Gaussian σ and reroll/splice/prune
    probabilities. With strength=0, offspring are near-identical to parents
    (only new id/name; numeric drift is zero, reroll probability is zero).
    """
    if not parents:
        raise ValueError("breed_generation requires at least 1 parent")

    s = max(0.0, min(1.0, strength))
    std = s * 0.45
    reroll_rate = s * 0.85

    if len(parents) == 1:
        return [_mutate(parents[0], std, reroll_rate) for _ in range(target_count)]

    offspring: list[Genome] = []
    for _ in range(target_count):
        idx_a = random.randrange(len(parents))
        idx_b = random.randrange(len(parents) - 1)
        if idx_b >= idx_a:
            idx_b += 1
        # Crossover already mixes from two parents; soften the post-mutation.
        child = _mutate(
            _crossover(parents[idx_a], parents[idx_b]),
            std * 0.5,
            reroll_rate * 0.5,
        )
        offspring.append(child)
    return offspring
